packages/connect/src/base/stun.ts

```diff
@@ -82,109 +82,98 @@ type Address = {
-  } else if (multiAddrs.length > DEFAULT_PARALLEL_STUN_CALLS) {
-    // Limit number of STUN servers to contact
-    usableMultiaddrs = randomSubset(multiAddrs, DEFAULT_PARALLEL_STUN_CALLS)
+  let selectedStunServers: Multiaddr[]
+  if (multiaddrs.length > DEFAULT_PARALLEL_STUN_CALLS) {
+    selectedStunServers = randomSubset(multiaddrs, DEFAULT_PARALLEL_STUN_CALLS)
   } else {
-    usableMultiaddrs = multiAddrs
+    selectedStunServers = multiaddrs
   }
 
-  verbose(`Trying to determine external IP by using ${usableMultiaddrs.map((m) => m.toString()).join(',')}`)
-
-  let responses = await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)
-
-  if (responses.length == 0 && (usingPublicServers || runningLocally)) {
-    // We have already contacted public and this did not lead to a result
-    // hence we cannot determine public IP address
-    return
-  }
-
-  if ([0, 1].includes(responses.length) && !runningLocally && !usingPublicServers) {
-    // Received too less results to see if addresses are ambiguous,
-    // let's try some public servers
-    usingPublicServers = true
-    usableMultiaddrs = randomSubset(PUBLIC_STUN_SERVERS, DEFAULT_PARALLEL_STUN_CALLS)
-    verbose(
-      `Using own STUN servers did not lead to a result. Trying ${usableMultiaddrs
-        .map((ma: Multiaddr) => ma.toString())
-        .join(', ')}`
-    )
+  let responses: RequestWithResponse[] = await performSTUNRequests(
+    selectedStunServers,
+    socket,
+    STUN_TIMEOUT,
+    runningLocally
+  )
+
+  if (multiaddrs.length > DEFAULT_PARALLEL_STUN_CALLS) {
+    while (responses.length < 2) {
+      for (const selected of selectedStunServers) {
+        usedStunServers.add(selected.toString())
+      }
 
-    responses.push(...(await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)))
-  }
+      if (usedStunServers.size >= maxAttempts) {
+        break
+      }
 
-  if (responses.length == 0) {
-    // Even contacting public STUN servers did not lead
-    return
-  }
+      const toFetch = Math.min(maxAttempts, DEFAULT_PARALLEL_STUN_CALLS + usedStunServers.size) - usedStunServers.size
 
-  let filteredResults = getUsableResults(responses, runningLocally)
+      selectedStunServers = randomSubset(multiaddrs, toFetch, (ma: Multiaddr) => !usedStunServers.has(ma.toString()))
 
-  if ([0, 1].includes(filteredResults.length) && !runningLocally && !usingPublicServers) {
-    // Received results were not usable to determine public IP address
-    // now trying public ones
-    verbose(
-      `Using own STUN servers did not lead to a result. Trying ${usableMultiaddrs
-        .map((ma: Multiaddr) => ma.toString())
-        .join(', ')}`
-    )
-    usingPublicServers = true
-    usableMultiaddrs = randomSubset(PUBLIC_STUN_SERVERS, DEFAULT_PARALLEL_STUN_CALLS)
-    responses.push(...(await performSTUNRequests(usableMultiaddrs, socket, STUN_TIMEOUT)))
```

Comment:
What if one of this `performSTUNRequests` throws ? Don't we still want to continue the loop to look for others?

Reply:
It does not. But I can change it

Reply:
fixed another error and performSTUNRequest should indeed not throw. Otherwise we could safely crash the node because that error is really important then.